### PR TITLE
[UI Tests] Make current WordPress UI Tests to run against Jetpack app

### PR DESCRIPTION
**Pull request by tiagomar**
### What
The objective of this PR is to make the current Wordpress UI Tests to run against the Jetpack app locally. 

### How
* `WordPress > WordPress UI Tests` folder was copied into `WordPress > Jetpack UI Tests`
* Creating `JetpackUITests` Test Plan

### Notes
* `MainNavigationTests` and `StatsTests` are disabled and will be enabled in an upcoming PR.
* Work to minimize duplications will be tackled in another PR after all the tests are enabled.

### Testing
<details>
  <summary>Run the <b>JetpackUITests</b> Test Plan and confirm all the enabled tests pass.</summary>
  image
</details>

### Regression Notes
No impacts are expected on the WordPress UI Tests as none of their files were changed.

### CODE REVIEW

**Comment by jostnes** [comment]
Thanks for working on this @tiagomar! Nice to see the tests finally running (and passing 🎉 ) on Jetpack app! 

image

One thing I'm unsure about this change is that would it be possible to create a separate test plan for the same scheme so we don't have to duplicate the files/folders (and in a state of having 2 sets of UI tests to maintain)? Not tested but it looks like it's possible based on this [article](https://useyourloaf.com/blog/xcode-test-plans/#:~:text=You%20can%20add%20multiple%20test,selection%20of%20tests%20and%20configurations.). (it looks like the difference is the `Target for Variable Expansion` property, if we could have test plans with different configurations I think we can set one to run on WP and another on JP, we could also control which tests are skipped on the `Tests` configuration)

I could be missing some of the rationales behind having the code separated, but do you think it's worth trying out creating the Jetpack test plan based on the `WordPressUITests` folder to maintain 1 UI test codebase?

**Comment by mokagio** [comment]
Thanks for this @tiagomar !

+1 on @jostnes' suggestion: It would be great to avoid wholesale duplicating the test code.

I had a quick look at using a test plan with different "Target for Variable Expansion" as suggested, but unfortunately that didn't work.

As you can see in the screenshot below, even though the target was Jetpack, the WordPress app was the one launched.

Screenshot 2023-02-02 at 4 43 57 pm

I believe this is due to UI tests targets making you requiring an hardcoded "Target Application":

image

I need to run now, but here are two things I'd look into next:

- The "Target Application" is set via the `TEST_TARGET_NAME` build setting. It could be possible to change the value when running the tests via the terminal
- We could use `XCUIApplication(bundleIdentifier: "com.automattic.jetpack")` (or `org.wordpress`) instead of `XCUIApplication()` to switch between the two apps from the same UI tests target. We could read which `bundleIdentifier` value to use from a test plan or scheme variable.

Thinking about it, though, I can see why we might want to have a dedicated UI test target for Jetpack in the future, as the app gets more and more features that WordPress doesn't have. 

If we decide to go down the dedicated target, then I would urge you to extract as much logic as possible away from the UI test target and into the UITestsFoundation one, so that it can be _shared_ from a single place instead of duplicated. The sharing could extend to flows as well as single screens, and should definitely include the contents of `[WordPress/JetpackUITests/Utils/XCTest+Extensions.swift](https://github.com/wordpress-mobile/WordPress-iOS/pull/20046/files#diff-f808c492d128547e4efc8af4e466e37d781cefc2564a404b4746089287b16c3b)`

**Comment by tiagomar** [comment]
Thank you @jostnes and @mokagio for taking a look!! 🙇 

> One thing I'm unsure about this change is that would it be possible to create a separate test plan for the same scheme so we don't have to duplicate the files/folders (and in a state of having 2 sets of UI tests to maintain)? 
>  -- @jostnes 

> +1 on @jostnes' suggestion: It would be great to avoid wholesale duplicating the test code.
>  -- @mokagio 

I could't agree more. I was concerned about the duplication but it might look much worse than what it should be in the end. More on this further down.

> Not tested but it looks like it's possible based on this [article](https://useyourloaf.com/blog/xcode-test-plans/#:~:text=You%20can%20add%20multiple%20test,selection%20of%20tests%20and%20configurations.). (it looks like the difference is the Target for Variable Expansion property, if we could have test plans with different configurations I think we can set one to run on WP and another on JP, we could also control which tests are skipped on the Tests configuration)
> -- @jostnes 

As we discussed, our case is not quite the same. The example you shared would work for running different sets of tests against the same app while what we need is to run different sets of tests but against different Target apps.

> I had a quick look at using a test plan with different "Target for Variable Expansion" as suggested, but unfortunately that didn't work.
>  -- @mokagio 

Yeah, that was my initial thought but I couldn't make it work. I tried replacing every `WordPress` reference I could find by 
`Jetpack` but faced the same issue as you, then I thought there was some hidden tying it to `WordPress` app. It only worked after I created a new UI Test Target with `Jetpack` as `Target to be Tested`.

I'm still trying a few alternatives, though, like 2 Test Plans pointing to a single Test Target, for example, but disabling one test in `WordPress` Test Plan also disables it in `Jetpack`. It's like modifying the actual Test Target instead of the Test Plans, which doesn’t work for us.

> Thinking about it, though, I can see why we might want to have a dedicated UI test target for Jetpack in the future, as the app gets more and more features that WordPress doesn't have.
>  -- @mokagio 

And this is why I started thinking that the duplication wouldn't be that bad in the end. The focus now is on `Jetpack` and not only `WordPress` will enter "maintenance mode" but it will also have features removed as part of the current project. All the Reader, Stats, Support Screen tests and others will no longer apply to `WordPress`, then removed from `WordPressUITests`.  

> If we decide to go down the dedicated target, then I would urge you to extract as much logic as possible away from the UI test target and into the UITestsFoundation one (...)
>  -- @mokagio 

Sure thing! Only the actual Tests would remain in the Test Targets. That was the idea but I initially thought on working on this right after having all the tests running in another PR, as mentioned in the Notes. In case we go that route, do you think I should have everything done in a single PR (this one) instead?

**Comment by mokagio** [comment]
@tiagomar 

> Only the actual Tests would remain in the Test Targets. That was the idea but I initially thought on working on this right after having all the tests running in another PR, as mentioned in the Notes. In case we go that route, do you think I should have everything done in a single PR (this one) instead?

Have you considered doing 1) extract all the shared logic (at least, all the logic you foresee being shared) in a PR, then 2) adding the new Jetpack UI tests target?

> as mentioned in the Notes

Ah! I missed the "Work to minimize duplications will be tackled in another PR after all the tests are enabled." Apologies.

I think if you consider this an experiment PR, it's fine to duplicate stuff. But I wouldn't recommend merging duplication only to remove it afterwards. Or, let me rephrase... the only case in which I'd think it'd be okay to do that instead of moving incrementally is if there is urgency in UI testing Jetpack in CI.

**Comment by tiagomar** [comment]
Hey, @mokagio!

> Have you considered doing 1) extract all the shared logic (at least, all the logic you foresee being shared) in a PR, then 2) adding the new Jetpack UI tests target?

I hadn't put much thought on this, to be honest. I just wanted to have the Jetpack tests running real soon without any impact on the WordPress side. But, yeah, I can see it definitely translating into rework and bad coding practices. 😅 

> I think if you consider this an experiment PR, it's fine to duplicate stuff. But I wouldn't recommend merging duplication only to remove it afterwards. Or, let me rephrase... the only case in which I'd think it'd be okay to do that instead of moving incrementally is if there is urgency in UI testing Jetpack in CI.

It makes total sense to me. It's not a urgency yet, so I'll put this PR back to draft and get it done the right way. Thank you very much! 🙇‍♂️

**Comment by tiagomar** [comment]
Closing this in favor of [20226](https://github.com/wordpress-mobile/WordPress-iOS/pull/20226).
